cmd/registry/patch/api.go

```diff
@@ -0,0 +1,196 @@
+	} `yaml:"data"`
+}
+
+func newAPI(ctx context.Context, client *gapic.RegistryClient, message *rpc.Api) (*API, error) {
+	apiName, err := names.ParseApi(message.Name)
+	if err != nil {
+		return nil, err
+	}
+	api := &API{
+		Header: Header{
+			APIVersion: REGISTRY_V1,
+			Kind:       "API",
+			Metadata: Metadata{
+				Name:        apiName.ApiID,
+				Labels:      message.Labels,
+				Annotations: message.Annotations,
+			},
+		},
+	}
+	api.Data.DisplayName = message.DisplayName
+	api.Data.Description = message.Description
+	api.Data.Availability = message.Availability
+	api.Data.RecommendedVersion, err = relativeVersionName(apiName, message.RecommendedVersion)
+	if err != nil {
+		return nil, err
+	}
+	api.Data.RecommendedDeployment, err = relativeDeploymentName(apiName, message.RecommendedDeployment)
+	if err != nil {
+		return nil, err
+	}
+	err = core.ListVersions(ctx, client, apiName.Version(""), "", func(message *rpc.ApiVersion) {
+		version, err2 := newAPIVersion(ctx, client, message)
+		// unset these because they can be inferred
+		version.APIVersion = ""
+		version.Kind = ""
+		if err2 == nil {
+			api.Data.APIVersions = append(api.Data.APIVersions, version)
+		} else {
+			err = err2
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	err = core.ListDeployments(ctx, client, apiName.Deployment(""), "", func(message *rpc.ApiDeployment) {
+		deployment, err2 := newAPIDeployment(ctx, client, message)
+		// unset these because they can be inferred
+		deployment.APIVersion = ""
+		deployment.Kind = ""
+		if err2 == nil {
+			api.Data.APIDeployments = append(api.Data.APIDeployments, deployment)
+		} else {
+			err = err2
+		}
+	})
+	return api, err
+}
+
+// ExportAPI allows an API to be individually exported as a YAML file.
+func ExportAPI(ctx context.Context, client *gapic.RegistryClient, message *rpc.Api) ([]byte, *Header, error) {
+	api, err := newAPI(ctx, client, message)
+	if err != nil {
+		return nil, nil, err
+	}
+	b, err := yaml.Marshal(api)
+	if err != nil {
+		return nil, nil, err
+	}
+	return b, &api.Header, nil
+}
+
+// relativeVersionName returns the version id if the version is within the specified API
+func relativeVersionName(apiName names.Api, version string) (string, error) {
+	if version == "" {
+		return "", nil
+	}
+	versionName, err := names.ParseVersion(version)
+	if err != nil {
+		return "", err
+	}
+	if versionName.Api().String() == apiName.String() {
+		return versionName.VersionID, nil
+	}
+	return version, nil
+}
+
+// relativeDeploymentName returns the deployment id if the deployment is within the specified API
+func relativeDeploymentName(apiName names.Api, deployment string) (string, error) {
+	if deployment == "" {
+		return "", nil
+	}
+	deploymentName, err := names.ParseDeployment(deployment)
+	if err != nil {
+		return "", err
+	}
+	if deploymentName.Api().String() == apiName.String() {
+		return deploymentName.DeploymentID, nil
+	}
+	return deployment, nil
+}
```

Comment:
These functions assume `version` is a valid name and fail the export if they aren't. Is this the intended behavior? 

More importantly, why are we modifying these fields during export? If the user exports an API that has a full resource name stored in these fields, the exported YAML will not accurately reflect the resource being exported. In practice this means that a user can run `registry export` on a resource then `registry apply` the result and end up with a different value in these fields.

IMO we should export resources without modification.

Reply:
This is intentional. Full (absolute) resource names are unwieldy and make it difficult to use YAML resource representations in multiple projects. The API itself follows the AIPs (which call for absolute path names), but I think these external forms should be designed to easily allow resources to be exported from one registry and reimported into another.

I've added a TODO to address the need for additional robustness and validation. I think it would be good for imports to accept both relative and absolute resource names and for exports to produce relative names.